### PR TITLE
feat: make group object class configurable

### LINKED ISSUE #48: group lookup: rename get_associtated_groups and return empty vector for no groups
`/Users/kasunranasinghe/Projects/Rust/simple-ldap/src/lib.rs:1468` has `get_associtated_groups` (typo) and returns `NotFound` when no groups exist.

## Problem
Typo in method name hurts API quality, and no-result currently raises `NotFound` where an empty list is often more ergonomic.

## Proposal
- Add a correctly spelled method (`get_associated_groups`) and mark `get_associtated_groups` as deprecated alias.
- Change behavior for zero matches to return `Ok(Vec::new())`.

## Acceptance criteria
- Preserve compatibility through deprecated typoed method.
- Add tests for zero-group user case and user-in-groups case.

### LINKED ISSUE #57: Allow providing an optional object class to get_associtated_groups
Groups don't necessarily have an `objectClass` of `groupOfNames` (in our AD it's just `groups` for example), so I would propose adding another parameter (`Option<&str>`) to be able to customise it.

If you'd like me to I could combine this with #48: create a new `get_associated_groups` method with the additional parameter and new return type, mark the old one as deprecated and just call the new one with `None` for the object class param. I would also keep the current return type for the deprecated method and just return an error when the search returns an empty result so that people can use the deprecated method as is until they switch.

### PR DESCRIPTION
**Pull request by aeveris**
This PR
* adds a new `get_associated_groups` method to `LdapClient` to correct a previous typo
* adds a new enum `GroupObjectClass`
* adds a new argument to `get_associated_groups` to let users specify a group's object class
* returns an empty `Vec` from `get_associated_groups` when no groups can be found
* deprecates `get_associtated_groups` and makes it a thin wrapper for `get_associated_groups`

closes #57, closes #48

### CODE REVIEW

**Comment by aeveris** [comment]
I've decided to go with the enum approach for now. I think adding more variants is easy enough should the need arise.

File: src/lib.rs [hunk 1]
```diff
@@ -176,6 +176,30 @@ pub extern crate ldap3;
...
+
+impl GroupObjectClass {
+    fn as_str(&self) -> &'static str {
+        match *self {
+            Self::Group => "group",
+            Self::GroupOfNames => "groupOfNames",
+            Self::GroupOfUniqueNames => "groupOfUniqueNames",
+        }
+    }
+}
```
Comment (Iizuki) [inline src/lib.rs]: Maybe implementing `Display` would be more appropriate here? It might be useful to someone outside as well, and as things currently stand we're not really avoiding Strings anyway.

File: src/lib.rs [hunk 2]
```diff
@@ -1385,18 +1410,20 @@ impl LdapClient {
...
+    ///         "uid=bd9b91ec-7a69-4166-bf67-cc7e553b2fd9,ou=people,dc=example,dc=com",
+    ///         None).await;
     /// }
     /// ```
-    pub async fn get_associtated_groups(
+    pub async fn get_associated_groups(
         &mut self,
         group_ou: &str,
         user_dn: &str,
+        grp_obj_cls: Option<GroupObjectClass>,
```
Comment (Iizuki) [inline src/lib.rs]: I wonder if `Option` here is the correct choice? We could just take the `GroupObjectClass` directly and document using the `Default` implementation, as that's what we're anyway doing behind the scenes. Option might lead to a misconception that we're not using anything when providing `None`.

File: src/lib.rs [hunk 3]
```diff
@@ -1483,7 +1560,7 @@ struct SerializeWrapper(#[serde(with = "Ldap3SearchEntry")] ldap3::SearchEntry);
 
 // Allowing users to debug serialization issues from the logs.
 #[instrument(level = Level::DEBUG)]
-fn to_signle_value<T: for<'a> Deserialize<'a>>(search_entry: SearchEntry) -> Result<T, Error> {
+fn to_single_value<T: for<'a> Deserialize<'a>>(search_entry: SearchEntry) -> Result<T, Error> {
```
Comment (Iizuki) [inline src/lib.rs:1569]: Always nice to fix typos ^^

**Review by Iizuki — CHANGES_REQUESTED** [review]
Thanks for the PR. I'll have a closer look when I get back to office in couple weeks, but I already threw there some ideas.

File: src/lib.rs (hunk 2 above)
Comment (keaz) [reply src/lib.rs]: yes, agree with this.

**Comment by Iizuki** [comment]
It seems we're in unison with regards to design questions. I think after fixing those this is good to merge @aeveris.

File: src/lib.rs [hunk 4]
```diff
@@ -1385,18 +1413,20 @@ impl LdapClient {
     ///
     ///     let mut client = LdapClient::new(ldap_config).await.unwrap();
     ///
-    ///     let result = client.get_associtated_groups("ou=groups,dc=example,dc=com",
-    ///     "uid=bd9b91ec-7a69-4166-bf67-cc7e553b2fd9,ou=people,dc=example,dc=com").await;
+    ///     let result = client.get_associated_groups("ou=groups,dc=example,dc=com",
+    ///         "uid=bd9b91ec-7a69-4166-bf67-cc7e553b2fd9,ou=people,dc=example,dc=com",
+    ///         GroupObjectClass::default()).await;
```
Comment (Iizuki) [inline src/lib.rs:1418]: This doc tests fails because of a missing import. You can run them with `cargo test --doc`.

File: src/lib.rs [hunk 5]
```diff
@@ -1385,18 +1413,20 @@ impl LdapClient {
...
     /// }
     /// ```
-    pub async fn get_associtated_groups(
+    pub async fn get_associated_groups(
         &mut self,
         group_ou: &str,
         user_dn: &str,
+        grp_obj_cls: GroupObjectClass,
```
Comment (Iizuki) [inline src/lib.rs]: I think the acronyms in the parameter name could be expanded for clarity. Remember the reference in docs above too.

**Review by Iizuki — CHANGES_REQUESTED** [review]
Little tweaks still

File: src/lib.rs (hunk 4 above)
Comment (aeveris) [reply src/lib.rs:1418]: Oops, sorry. Fixed

File: src/lib.rs (hunk 5 above)
Comment (aeveris) [reply src/lib.rs]: Sure, makes sense. Done

**Comment by Iizuki** [comment]
Thanks. I think it's all good now, but I encountered some unrelated issues when trying to run the tests. I submitted a fix in #63. I'd still ask you to rebase this after that gets merged so we can get the tests running.

**Comment by keaz** [comment]
@aeveris , merged the #63 you can rebase.

**Comment by aeveris** [comment]
> @aeveris , merged the #63 you can rebase.

Done

**Review by Iizuki — APPROVED** [review]

**Comment by Iizuki** [comment]
Thanks for your contribution @aeveris! It's merged and released now.
